kroxylicious-schema-tools/kroxylicious-schema-tool/src/main/java/io/kroxylicious/tools/schema/compiler/CodeGen.java

```diff
@@ -0,0 +1,713 @@
+    private final String nonNullAnnotation;
+    private final List<TypeAnnotator> typeAnnotators;
+    private final List<PropertyAnnotator> propertyAnnotators;
+
+    public CodeGen(Diagnostics diagnostics,
+                   IdVisitor idVisitor,
+                   Map<String, String> existingClasses,
+                   String nullableAnnotation,
+                   String nonNullAnnotation,
+                   List<TypeAnnotator> typeAnnotators,
+                   List<PropertyAnnotator> propertyAnnotators) {
+        this.diagnostics = Objects.requireNonNull(diagnostics);
+        this.idVisitor = Objects.requireNonNull(idVisitor);
+        this.existingClasses = existingClasses;
+        this.nullableAnnotation = nullableAnnotation;
+        this.nonNullAnnotation = nonNullAnnotation;
+        this.typeAnnotators = typeAnnotators;
+        this.propertyAnnotators = propertyAnnotators;
+    }
+
+    SchemaObject resolveRef(SchemaObject root, SchemaObject schema) {
+        var ref = schema.getRef() == null ? null : URI.create(schema.getRef());
+        if (ref != null) {
+            if (ref.isAbsolute()) {
+                diagnostics.reportFatal("Use of an absolute URI in $ref is not supported");
+                return new SchemaObject();
+            }
+            else if ((ref.getPath() == null
+                    || ref.getPath().isEmpty())
+                    && ref.getFragment() != null) {
+                return resolveInternalFragmentRef(root, ref);
+            }
+            // Two possibilities: a local ref point to a file which we should be compiling
+            // or a local ref pointing to a file which we should _not be compiling_
+            // In the _not compiling_ case the referring file depends on decls which we won't emit
+            // If the java for the referred to file is out of date then we'll likely generate the wrong thing
+
+            URI sought = URI.create(root.getId()).resolve(ref);
+            var resolved = idVisitor.resolve(sought);
+            if (resolved != null) {
+                return resolved;
+            }
+            else {
+                diagnostics.reportError("Canot resolve $ref (but $ref not fully supported) {}", ref);
+                return new SchemaObject();
+            }
+        }
+        else {
+            return schema;
+        }
+    }
+
+    @NonNull
+    private SchemaObject resolveInternalFragmentRef(SchemaObject root, URI ref) {
+        if (ref.getFragment().startsWith("/definitions/")) {
+            Map<String, SchemaObject> defs = root.getDefinitions();
+            if (defs != null) {
+                String name = ref.getFragment().substring("/definitions/".length());
+                SchemaObject object = defs.get(name);
+                if (object != null) {
+                    return new SchemaObjectBuilder(object).withJavaType(name).build();
+                }
+            }
+            diagnostics.reportFatal("Couldn't resolve $ref " + ref);
```

Comment:
Why is this fail fast? 

Is it because if we can't resolve a ref all bets are off for subsequent generation?